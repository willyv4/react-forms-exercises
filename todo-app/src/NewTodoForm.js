import { useState } from "react";

const NewTodoForm = ({ addTodo }) => {
  const INITIAL_STATE = { todo: "" };
  const [formData, setFormData] = useState(INITIAL_STATE);

  const handleSubmit = (evt) => {
    evt.preventDefault();
    addTodo(formData);
    setFormData(INITIAL_STATE);
  };

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="todo">Todo:</label>
      <input
        id="todo"
        name="todo"
        type="text"
        value={formData.todo}
        onChange={handleChange}
      />
      <button>Add Todo</button>
    </form>
  );
};
export default NewTodoForm;
