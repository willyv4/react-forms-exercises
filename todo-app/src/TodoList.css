.todo-list {
	
	width: 200px;
	
}