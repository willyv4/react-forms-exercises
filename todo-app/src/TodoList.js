import NewTodoForm from "./NewTodoForm";
import { useState } from "react";
import { v4 as uuid } from "uuid";
import "./TodoList.css";

const TodoList = () => {
  const [todo, setTodo] = useState([]);

  const addTodo = (todo) => {
    let newTodo = { ...todo, id: uuid() };
    setTodo((todo) => [...todo, newTodo]);
  };

  const removeTodo = (id) => {
    const currentBoxes = todo.filter((t) => t.id !== id);
    setTodo(currentBoxes);
  };

  const handleRemove = (id) => {
    removeTodo(id);
  };

  const renderTodos = () => {
    return (
      <ul>
        {todo.map((todo) => (
          <div key={todo.id}>
            <li>{todo.todo}</li>
            <button onClick={() => handleRemove(todo.id)}>X</button>
          </div>
        ))}
      </ul>
    );
  };

  return (
    <div className="todo-list">
      <NewTodoForm addTodo={addTodo} />
      {renderTodos()}
    </div>
  );
};
export default TodoList;
