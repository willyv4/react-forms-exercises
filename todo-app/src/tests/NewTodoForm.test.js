import NewTodoForm from "../NewTodoForm";
import { render, fireEvent, screen } from "@testing-library/react";
import TodoList from "../TodoList";

it("renders without crashing", function () {
  render(<NewTodoForm />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<NewTodoForm />);
  expect(asFragment()).toMatchSnapshot();
});

test("NewTodoForm submits new box data", () => {
  const addTodo = jest.fn();

  render(<NewTodoForm addTodo={addTodo} />);

  fireEvent.change(screen.getByLabelText("Todo:"), {
    target: { value: "go Sk8" },
  });

  fireEvent.click(screen.getByText("Add Todo"));

  expect(addTodo).toHaveBeenCalledWith({
    todo: "go Sk8",
  });
});

test("added todo can be deleted", () => {
  const addTodo = jest.fn();

  render(<TodoList addTodo={addTodo} />);

  fireEvent.change(screen.getByLabelText("Todo:"), {
    target: { value: "go Sk8 2.0" },
  });

  fireEvent.click(screen.getByText("Add Todo"));

  expect(screen.getByText("go Sk8 2.0")).toBeInTheDocument();

  fireEvent.click(screen.getByText("X"));

  expect(screen.queryByText("go Sk8 2.0")).toBeNull();
});
