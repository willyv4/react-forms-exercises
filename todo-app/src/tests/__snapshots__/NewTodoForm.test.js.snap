// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="todo"
    >
      Todo:
    </label>
    <input
      id="todo"
      name="todo"
      type="text"
      value=""
    />
    <button>
      Add Todo
    </button>
  </form>
</DocumentFragment>
`;
